# Movies Rating Prediction

Goal: Bulid the model that predicts the rating of a movie.

### 1. Import LIbraries and Dataset

In [11]:
import pandas as pd
import numpy as np

Loading the dataset into DataFrame

In [9]:
movie = pd.read_csv("C:/Data Science/CodSoft_Internship/Movie_Rating_Prediction_with_Python/IMDb Movies India.csv", encoding = 'latin1')

In [10]:
movie.head()

Name    Year Duration            Genre  \
0                                         NaN      NaN            Drama   
1  #Gadhvi (He thought he was Gandhi)  (2019)  109 min            Drama   
2                         #Homecoming  (2021)   90 min   Drama, Musical   
3                             #Yaaram  (2019)  110 min  Comedy, Romance   
4                   ...And Once Again  (2010)  105 min            Drama   

   Rating Votes            Director       Actor 1             Actor 2  \
0     NaN   NaN       J.S. Randhawa      Manmauji              Birbal   
1     7.0     8       Gaurav Bakshi  Rasika Dugal      Vivek Ghamande   
2     NaN   NaN  Soumyajit Majumdar  Sayani Gupta   Plabita Borthakur   
3     4.4    35          Ovais Khan       Prateik          Ishita Raj   
4     NaN   NaN        Amol Palekar  Rajat Kapoor  Rituparna Sengupta   

           Actor 3  
0  Rajendra Bhatia  
1    Arvind Jangid  
2       Roy Angana  
3  Siddhant Kapoor  
4      Antara Mali

Dimensions of dataset:

In [14]:
movie.shape

(15509, 10)

There are 15509 rows and 10 columns

In [15]:
# Summary of the DataFrame
movie.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15509 entries, 0 to 15508
Data columns (total 10 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   Name      15509 non-null  object 
 1   Year      14981 non-null  object 
 2   Duration  7240 non-null   object 
 3   Genre     13632 non-null  object 
 4   Rating    7919 non-null   float64
 5   Votes     7920 non-null   object 
 6   Director  14984 non-null  object 
 7   Actor 1   13892 non-null  object 
 8   Actor 2   13125 non-null  object 
 9   Actor 3   12365 non-null  object 
dtypes: float64(1), object(9)
memory usage: 1.2+ MB


In [16]:
# Displaying discriptive statistics of numerical columns
movie.describe()

Rating
count  7919.000000
mean      5.841621
std       1.381777
min       1.100000
25%       4.900000
50%       6.000000
75%       6.800000
max      10.000000

In [17]:
movie.describe(include='all')

Name    Year Duration  Genre       Rating Votes      Director  \
count    15509   14981     7240  13632  7919.000000  7920         14984   
unique   13838     102      182    485          NaN  2034          5938   
top     Anjaam  (2019)  120 min  Drama          NaN     8  Jayant Desai   
freq         7     410      240   2780          NaN   227            58   
mean       NaN     NaN      NaN    NaN     5.841621   NaN           NaN   
std        NaN     NaN      NaN    NaN     1.381777   NaN           NaN   
min        NaN     NaN      NaN    NaN     1.100000   NaN           NaN   
25%        NaN     NaN      NaN    NaN     4.900000   NaN           NaN   
50%        NaN     NaN      NaN    NaN     6.000000   NaN           NaN   
75%        NaN     NaN      NaN    NaN     6.800000   NaN           NaN   
max        NaN     NaN      NaN    NaN    10.000000   NaN           NaN   

            Actor 1 Actor 2 Actor 3  
count         13892   13125   12365  
unique         4718    4891    4820  
top     Ashok Kumar   Rekha    Pran  
freq            158      83      91  
mean            NaN     NaN     NaN  
std             NaN     NaN     NaN  
min             NaN     NaN     NaN  
25%             NaN     NaN     NaN  
50%             NaN     NaN     NaN  
75%             NaN     NaN     NaN  
max             NaN     NaN     NaN

### 2. Exploratory Data Analysis

Checking the missing data

In [18]:
movie.isnull()

Name   Year  Duration  Genre  Rating  Votes  Director  Actor 1  \
0      False   True      True  False    True   True     False    False   
1      False  False     False  False   False  False     False    False   
2      False  False     False  False    True   True     False    False   
3      False  False     False  False   False  False     False    False   
4      False  False     False  False    True   True     False    False   
...      ...    ...       ...    ...     ...    ...       ...      ...   
15504  False  False      True  False   False  False     False    False   
15505  False  False     False  False   False  False     False    False   
15506  False  False      True  False    True   True     False    False   
15507  False  False      True  False    True   True      True     True   
15508  False  False     False  False   False  False     False    False   

       Actor 2  Actor 3  
0        False    False  
1        False    False  
2        False    False  
3        False    False  
4        False    False  
...        ...      ...  
15504    False    False  
15505    False    False  
15506     True     True  
15507     True     True  
15508    False    False  

[15509 rows x 10 columns]

In [19]:
movie.isnull().sum()

Name           0
Year         528
Duration    8269
Genre       1877
Rating      7590
Votes       7589
Director     525
Actor 1     1617
Actor 2     2384
Actor 3     3144
dtype: int64

Here we can see that, 

column 'Year' has 528 missing values, 

column 'Duration' has 8296 missing values, 

column 'Genre' has 1877 missing values, 

column 'Rating' has 7590 missing values, 

column 'Votes' has 7589 missing values,

column 'Director' has 525 missing values, 

column 'Actor 1' has 1617 missing values, 

column 'Actor 2' has 2384 missing values 

and column 'Actor 3' has 3144 missing values. 

#### Remove missing values

In [32]:
movie.dropna(subset=['Rating'], inplace = True)
(round(movie.isnull().sum()/movie.isnull().count(), 4)*100).sort_values(ascending=False)

Duration    26.11
Actor 3      3.69
Actor 2      2.53
Actor 1      1.58
Genre        1.29
Director     0.06
Name         0.00
Year         0.00
Rating       0.00
Votes        0.00
dtype: float64

The missing percentage in other variables is not significant except for the variable duration so let's remove missing values from these variables.

In [33]:
movie.dropna(subset=['Director', 'Actor 1', 'Actor 2', 'Actor 3','Genre'], inplace = True)
(round(movie.isnull().sum()/movie.isnull().count(),4)*100).sort_values(ascending=False)

Duration    25.13
Name         0.00
Year         0.00
Genre        0.00
Rating       0.00
Votes        0.00
Director     0.00
Actor 1      0.00
Actor 2      0.00
Actor 3      0.00
dtype: float64

Now lets impute missing values in duration to the mean duration value.

In [34]:
print(movie['Duration'].dtype)
print(movie['Duration'].head())


object
1    109 min
3    110 min
5    147 min
6    142 min
8     82 min
Name: Duration, dtype: object


In [37]:
# Convert column to string and remove 'min'
movie['Duration'] = movie['Duration'].astype(str).str.replace('min', '', regex=True)

# Convert to numeric, setting errors='coerce' to handle any invalid cases
movie['Duration'] = pd.to_numeric(movie['Duration'], errors='coerce')

# Fill NaN values with mean
movie['Duration'].fillna(movie['Duration'].mean(), inplace=True)

# Check for null values
print(movie.isnull().sum())


Name        0
Year        0
Duration    0
Genre       0
Rating      0
Votes       0
Director    0
Actor 1     0
Actor 2     0
Actor 3     0
dtype: int64


In [38]:
print(movie['Duration'].dtype)  # Should be int64 or float64
print(movie.head())  # Verify values are cleaned and numeric


float64
                                 Name    Year  Duration  \
1  #Gadhvi (He thought he was Gandhi)  (2019)     109.0   
3                             #Yaaram  (2019)     110.0   
5                ...Aur Pyaar Ho Gaya  (1997)     147.0   
6                           ...Yahaan  (2005)     142.0   
8                  ?: A Question Mark  (2012)      82.0   

                       Genre  Rating  Votes        Director          Actor 1  \
1                      Drama     7.0      8   Gaurav Bakshi     Rasika Dugal   
3            Comedy, Romance     4.4     35      Ovais Khan          Prateik   
5     Comedy, Drama, Musical     4.7    827    Rahul Rawail       Bobby Deol   
6        Drama, Romance, War     7.4  1,086  Shoojit Sircar  Jimmy Sheirgill   
8  Horror, Mystery, Thriller     5.6    326   Allyson Patel        Yash Dave   

                  Actor 2          Actor 3  
1          Vivek Ghamande    Arvind Jangid  
3              Ishita Raj  Siddhant Kapoor  
5  Aishwarya Rai Bach

In [39]:
movie.drop_duplicates(inplace = True)

In [40]:
df.shape

(7558, 10)